walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,97 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
 
 public final class FindMeetingQuery {
+
+  /** General algorithm (first attempt, could be better optimized, generally runs in linear
+   * time except for calls to Collections.disjoint() in the getAllRelevantEvents() method): 
+   * 1. Get a list of all of the times when attendees are busy.
+   * 2. Sort the list by start time.
+   * 3. Loop through, and find all available times "between" these events; requires some amount
+   * of shifting around bounds to deal with edge cases regarding nested events.
+   * 4. Return a list of these times, only adding a time if it is longer than the proposed
+   * duration of the MeetingRequest. 
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  	Collection<TimeRange> relevantTimes = getAllRelevantTimes(events, request);
+    List<TimeRange> sortedTimes = getSortedTimeRanges(relevantTimes);
+    return getNonOverlappingTimes(sortedTimes, request);
+  }
+
+  /**
+   * Returns a Collection of only those TimeRanges which correspond to events
+   * with attendees common to the 'request' parameter.
+   */
+  private Collection<TimeRange> getAllRelevantTimes(Collection<Event> events, MeetingRequest request) {
+    Collection<String> attendees = request.getAttendees();
+    Collection<TimeRange> relevantTimes = new HashSet<>(); // Prevent duplicate times.
+    for (Event event : events) {
+      Collection<String> eventAttendees = event.getAttendees();
+      if (!Collections.disjoint(attendees, eventAttendees)) {
+        relevantTimes.add(event.getWhen());
+      }
+    }
+    return relevantTimes;
+  }
+
+  /**
+  * Returns a sorted (by start time) List<TimeRange> representing the Collection of events
+  * passed as a parameter. It is necessary to convert to a list for use of the sort() method.
+  */
+  private List<TimeRange> getSortedTimeRanges(Collection<TimeRange> times) {
+    List<TimeRange> sortedTimes = new ArrayList<>(times);
+    Collections.sort(sortedTimes, TimeRange.ORDER_BY_START);
+    return sortedTimes;
+  }
+
+  /** Returns a Collection of only those times which do not overlap with any TimeRange in 'times' */
```

Comment:
nit: use {@code times} when referring to the param.

Reply:
Also nit. Use a period at the end of sentences